### PR TITLE
report only modified options in reportActionDone(closes #4926)

### LINKED ISSUE #4926: Improve reporting action options in onTestActionStart/Done

### PR DESCRIPTION
**Pull request by AlexKamaev**
We pass the `command` object to the `action-start(done)` methods.
Some commands have `options`.
Options have default values.
User can pass options to action.

I.e. ClickCommand has ClickOptions:
```JS
{
    modifiers: {
        ctrl: Boolean,
        alt: Boolean,
        shift: Boolean,
        meta: Boolean
    },

    offsetX: Number,
    offsetY: Number,
    caretPos: Number,
    speed: Number
}
```
User executes action: `await t.click('#target'), { offsetX: 1, modifiers: { ctrl: true, alt: false } }`

The options will be the following:
```JS
{
    speed: null
    offsetX: 1
    offsetY: null
    modifiers: {
        ctrl: true
        alt: false
        shift: false
        meta: false
    }
    caretPos: null
}
```

We need to pass to reporter only NOT DEFAULT VALUES: `{ offsetX: 1, modifiers: { ctrl: true  } }`.

**P.S.**
I added the missing `timeout` option for selectors

### CODE REVIEW

File: src/utils/diff.ts [hunk 1]
```diff
@@ -0,0 +1,54 @@
+import { set, isEmpty, isEqual, isNil as isNullOrUndefined } from 'lodash';
+import { Dictionary } from '../configuration/interfaces';
+
+const OBJECTS_EMPTY_ERROR                = 'objects should be not empty';
```
Comment (miherlosev) [inline src/utils/diff.ts:4]: Error messages should start from an uppercase letter.

File: src/utils/diff.ts [hunk 2]
```diff
@@ -0,0 +1,54 @@
...
+function getFullPropertyPath (property: string, parentProperty: string): string {
+    if (parentProperty)
+        return `${parentProperty}.${property}`;
+
+    return property;
+}
+
+function validate (source: Dictionary<object>, modified: Dictionary<object>): void {
+    if (isEmpty(source) || isEmpty(modified))
+        throw new TypeError(OBJECTS_EMPTY_ERROR);
```
Comment (miherlosev) [inline src/utils/diff.ts:16]: We use the `GeneralError` class for argument validation errors.
* https://github.com/DevExpress/testcafe/blob/master/src/index.js#L19
* https://github.com/DevExpress/testcafe/blob/master/src/index.js#L32
and so on.

File: test/server/test-controller-events-test.js [hunk 3]
```diff
@@ -225,4 +240,77 @@ describe('TestController action events', () => {
                 expect(resultDuration).to.be.a('number').with.above(0);
             });
     });
+
+    it('Default command options should not be passed to the `reportTestActionDone` method', async () => {
+        const doneLog  = [];
```
Comment (miherlosev) [inline test/server/test-controller-events-test.js:245]: Let's rename as `actualLog` or just `log`.

File: test/server/util-test.js [hunk 4]
```diff
@@ -39,6 +41,29 @@ describe('Utils', () => {
         expect(escapeUserAgent('Chrome 67.0.3396 / Windows 8.1.0.0')).eql('Chrome_67.0.3396_Windows_8.1.0.0');
     });
 
+    it('Diff', () => {
+        expect(diff(null, null)).is.null;
```
Comment (miherlosev) [inline test/server/util-test.js:45]: There are no server tests for checking argument  validation errors.

File: src/utils/diff.ts (hunk 1 above)
Comment (AlexKamaev) [reply src/utils/diff.ts:4]: These errors are not TestCafe errors. 
We need them to immediately stop reccursive function on the deepest level.
I think, that this options comparison should never throw error in TestCafe, so we do not need `GeneralError` as well.
If the comparison is failed for some reason I just plan to catch the error and return null. However, in the current implementation compared objects are always of the same type and not empty. I added this validation to save us from future changes.

File: test/server/util-test.js (hunk 4 above)
Comment (AlexKamaev) [reply test/server/util-test.js:45]: I expect that result of some tests is null.
In this case using the throw/catch is just a way to stop function execution on the deep reccursion level.

File: src/reporter/command/command-formatter.ts [hunk 5]
```diff
@@ -91,12 +108,18 @@ export class CommandFormatter {
...
+            const property = this._command[key];
 
-            if (prop instanceof ExecuteSelectorCommand)
-                formattedCommand[key] = this._prepareSelector(prop, key);
+            if (property instanceof ExecuteSelectorCommand)
+                formattedCommand[key] = this._prepareSelector(property, key);
+            else if (isCommandOptions(property)) {
+                const modifiedOptions = CommandFormatter._getModifiedOptions(property);
+
+                if (modifiedOptions)
```
Comment (miherlosev) [inline src/reporter/command/command-formatter.ts:118]:
```js
if (!isEmpty(modifiedOptions))
```

**Pull request merged**